src/main/java/org/broadinstitute/hellbender/utils/codecs/GENCODE/GencodeGtfCodec.java

```diff
@@ -0,0 +1,376 @@
+
+        GencodeGtfFeature decodedFeature = null;
+
+        // Create some caches for our data (as we need to group it):
+        GencodeGtfGeneFeature gene = null;
+        GencodeGtfTranscriptFeature transcript = null;
+        final ArrayList< GencodeGtfExonFeature > exonStore = new ArrayList<>();
+        final ArrayList< GencodeGtfFeature > leafFeatureStore = new ArrayList<>();
+
+        // Accumulate lines until we have a full gene and all of its internal features:
+        while ( lineIterator.hasNext() ) {
+
+            String line = lineIterator.peek();
+
+            // We must assume we can get header lines.
+            // If we get a header line, we return null.
+            // This allows indexing to work.
+            if ( line.startsWith(COMMENT_START) ) {
+                lineIterator.next();
+                return null;
+            }
+
+            String[] splitLine = line.split("\t", -1);
+
+            // Ensure the file is at least trivially well-formed:
+            if (splitLine.length != NUM_COLUMNS) {
+                throw new UserException.MalformedFile("Found an invalid number of columns in the given GENCODE file on line "
+                        + currentLineNum + " - Given: " + splitLine.length + " Expected: " + NUM_COLUMNS + " : " + line);
+            }
+
+            // We need to key off the feature type to collapse our accumulated records:
+            final String featureType = splitLine[2];
+
+            // Create a baseline feature to add into our data:
+            GencodeGtfFeature feature = GencodeGtfFeature.create(splitLine);
+
+            // Make sure we keep track of the line number for if and when we need to write the file back out:
+            feature.setFeatureOrderNumber(currentLineNum);
+
+            // Once we see another gene, we take all accumulated records and combine them into a
+            // GencodeGtfFeature.
+            if ((gene != null) && (transcript != null) && (featureType.equals("gene") || featureType.equals("transcript") )) {
+
+                aggregateRecordsIntoGeneFeature(gene, transcript, exonStore, leafFeatureStore);
+
+                if ( featureType.equals("gene") ) {
+                    decodedFeature = gene;
+                    break;
+                }
+                else if ( featureType.equals("transcript") ) {
+                    transcript = (GencodeGtfTranscriptFeature) feature;
+                    ++currentLineNum;
+                }
+            }
+            else {
+                // We have not reached the end of this set of gene / transcript records.
+                // We must cache these records together so we can create a meaningful data hierarchy from them all.
+                // Records are stored in their Feature form, not string form.
+
+                // Add the feature to the correct storage unit for easy assembly later:
+                switch (featureType) {
+                    case "gene":
```

Comment:
There is actually already a FeatureType enum in GencodeGtfFeature and a getter for it.  I'll update the code to cue off of these.